Remove cloud-init from the Openstack image configuration

nlewo opened the pull request:
- `cloud-init` is replaced by modules used by our EC2 image (Openstack metadata service provides an EC2 API).
- Tests are added: check if `ssh` keys and `configuration.nix` are applied
- `novaImage` is renamed to `openstackImage`

This is mainly to reduce the size of the image (from 2.2GB to 1.3GB). Also, I think we generally don't care about `cloud-init` features since we have a `configuration.nix` file.

Unfortunately, one of the two tests doesn't work in the sandbox because it needs Internet. This is also the case of EC2 tests... I hope to fix them later (help is welcome on that).
I also test this image on an Openstack cloud. It is working as expected.

Note this also fixes the current Nova image which is currently broken.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

nlewo committed:
nova-image: use wget instead of cloud-init (via EC2 API)

The Openstack metadata service exposes the EC2 API. We use the
existing `ec2.nix` module to configure the hostname and ssh keys of an
Openstack Instance.

A test checks the ssh server is well configured.

This is mainly to reduce the size of the image (700MB). Also,
declarative features provided by cloud-init are not really useful
since we would prefer to use our `configuration.nix` file instead.

nlewo committed:
nova-image: add amazon-init module to the nova image

This allows the VM to provide a `configuration.nix` file to the VM.

The test doesn't work in sandbox because it needs Internet (however it
works interactively).

nlewo committed:
Rename `novaImage` to `openstackImage`

People don't necessary know `nova` is related to Openstack (it is a
component of Openstack). So, it is more explicit to call it
`openstackImage`.

jonas-chevalier-da commented on nixos/modules/virtualisation/openstack-config.nix:
```diff
@@ -0,0 +1,69 @@
+      after = [ "network-online.target" ];
+      script =
+        ''
+          metaDir=/etc/ec2-metadata
```
shouldn't this be under `/var` ? I don't know who consumes that folder data.

nlewo committed:
nixos/openstackImage: default hostname is empty string

This is to let the `ec2-data.nix` module sets the hostname from the
metadata API value.

nlewo commented:
@GrahamcOfBorg test openstack-image-metadata

nlewo commented:
@GrahamcOfBorg test ec2-nixops

nlewo commented:
The ec2 tests failure is not related to these patch (same error without my patches).

ping @flokli :)

flokli commented on nixos/modules/virtualisation/openstack-config.nix:
```diff
@@ -0,0 +1,72 @@
+    # Force getting the hostname from Nova.
+    networking.hostName = mkDefault "";
+
+    systemd.services.nova-init = {
```
this should probably be renamed

flokli commented on nixos/modules/virtualisation/openstack-config.nix:
```diff
@@ -0,0 +1,72 @@
+            wget --retry-connrefused -O "$metaDir/user-data" http://169.254.169.254/1.0/user-data && chmod 600 "$metaDir/user-data"
+          fi
+
+          if ! [ -e "$metaDir/hostname" ]; then
```
this should be optional, like in `nixos/modules/virtualisation/ec2-data.nix`

nlewo replied:
I would prefer to avoid changes on the ec2 image. I can't test the change (no ec2 account and the test is broken). Moreover this could also impact users relying on this `hostname` file.
I think this could be addressed in another PR.

flokli commented on nixos/modules/virtualisation/openstack-config.nix:
```diff
@@ -0,0 +1,72 @@
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+      script =
+        ''
```
The following script looks pretty similar to what's in nixos/modules/virtualisation/amazon-image.nix at `boot.initrd.postMountCommands`.

This duplicates the script in nixos/modules/virtualisation/amazon-image.nix, and probably should be merged.

Other suggestion, what about using https://github.com/coreos/coreos-metadata, which already handles changing hostname and ssh keys, and where it should be fairly trivial to make it expose user-data?

nlewo replied:
As discussed, I've moved the script in a shared file.

nlewo committed:
nixos: Add ec2-metadata-fetcher.nix file

To share the metadata fetcher script between ec2 and Openstack images.

flokli commented:
@edolstra from your reaction on the PR, is there a way you can test this?

Mic92 commented on nixos/maintainers/scripts/openstack/openstack-image.nix:
```diff
@@ -7,10 +7,10 @@ with lib;
     ];
 
-  system.build.novaImage = import ../../../lib/make-disk-image.nix {
+  system.build.openstackImage = import ../../../lib/make-disk-image.nix {
```
@Lassulus this might need a fix in https://github.com/nix-community/nixos-generators too

nlewo replied:
I could do it.
I also plan to send a post on discourse to ask for some help to publish this image on the NixOS download page.

nlewo committed:
nixos/tests/ec2: reuse ssh keys from ssh-keys.nix
